### PR TITLE
Bug fixes in SESE loop canonicalization for nested loops 

### PR DESCRIPTION
**Pull request by bgogul**
This patch contains the following changes
  - Avoid invalidating the use_iterator when updating use of escaping values.
  - Recompute dominator information after every loop change as it stale.

This effectively takes care of SESE loop canonicalization. 

Resolves [SR-7765](https://bugs.swift.org/browse/SR-7765).

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: test/TensorFlowRuntime/sese_loop_canonicalization.swift [hunk 1]
```diff
@@ -45,4 +45,51 @@ ControlFlowTests.testAllBackends("natSumWithBreak") {
...
+func sumOfProductsWithBound(_ M : Int32, _ N : Int32, _ Bound : Float) -> Tensor<Float> {
+  // Effectively computes natSum(M)*natSum(N) upto Bound
+  var sum = Tensor<Float>(0)
+outer_loop:for i in 0..<M {
+    for j in 0..<N {
+      // i + 1 and j + 1 is a workaround for
+      // https://bugs.swift.org/browse/SR-8256
+      let prod = Tensor<Float>(Float((i + 1) * (j + 1)));
+      sum += prod
+      if (sum.scalarized() >= Bound) {
```
Comment (rxwei) [inline test/TensorFlowRuntime/sese_loop_canonicalization.swift]: No need for semicolons or parens around conditions. This happens to me too when I write too much C++ :)

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 2]
```diff
@@ -461,7 +461,13 @@ SingleExitLoopTransformer::createNewHeader() {
     SILValue newValue = newHeader->createPHIArgument(
         escapingValue->getType(), escapingValue.getOwnershipKind());
     // Replace uses *outside* of the loop with the new value.
-    for (Operand *use : escapingValue->getUses()) {
+    ValueBase::use_iterator UI = escapingValue->use_begin(),
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Use `auto`? We use `auto` quite extensively in this code base.

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Review by lattner — APPROVED** [review]
Agree that auto would be better there.

**Pull request merged**